src/routers/middleware_router.rs

```diff
@@ -53,15 +51,41 @@ impl Router<(FunctionInfo, HashMap<String, String>), MiddlewareRoute> for Middle
 
 impl MiddlewareRouter {
     pub fn new() -> Self {
+        let mut globals = HashMap::new();
+        globals.insert(MiddlewareType::BeforeRequest, RwLock::new(vec![]));
+        globals.insert(MiddlewareType::AfterRequest, RwLock::new(vec![]));
         let mut routes = HashMap::new();
         routes.insert(
-            MiddlewareRoute::BeforeRequest,
+            MiddlewareType::BeforeRequest,
             RwLock::new(MatchItRouter::new()),
         );
         routes.insert(
-            MiddlewareRoute::AfterRequest,
+            MiddlewareType::AfterRequest,
             RwLock::new(MatchItRouter::new()),
         );
-        Self { routes }
+        Self { globals, routes }
+    }
+
+    pub fn add_global_middleware(
+        &self,
+        middleware_type: &MiddlewareType,
+        function: FunctionInfo,
+    ) -> Result<()> {
+        self.globals
+            .get(middleware_type)
+            .context("No relevant map")?
+            .write()
+            .unwrap()
```

Comment:
can we use `?` instead of `unwrap` here?

Reply:
I still have to figure out a way to do that, in this case to my understanding we locked the data so it is not thread safe anymore, and we cannot return the error safely here. The compiler won't let me use `?` in this case.

Reply:
Oh, okay. Got it